core/client/src/main/java/alluxio/client/file/FileSystemContext.java

```diff
@@ -90,60 +231,165 @@ public void close() {
+      BlockWorkerThriftClientPool pool = new BlockWorkerThriftClientPool(mParentSubject, rpcAddress,
+          Configuration.getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_SIZE_MAX),
+          Configuration.getLong(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_GC_THRESHOLD_MS));
+      if (mBlockWorkerClientPools.putIfAbsent(rpcAddress, pool) != null) {
+        pool.close();
+      }
+    }
+
+    if (!mBlockWorkerHeartbeatClientPools.containsKey(rpcAddress)) {
+      BlockWorkerThriftClientPool pool = new BlockWorkerThriftClientPool(mParentSubject, rpcAddress,
+          Configuration.getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_SIZE_MAX),
+          Configuration.getLong(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_GC_THRESHOLD_MS));
+      if (mBlockWorkerHeartbeatClientPools.putIfAbsent(rpcAddress, pool) != null) {
+        pool.close();
+      }
+    }
+
+    return new RetryHandlingBlockWorkerClient(mBlockWorkerClientPools.get(rpcAddress),
+        mBlockWorkerHeartbeatClientPools.get(rpcAddress), address, sessionId);
+  }
+
+  /**
    * Creates a new file system worker client, prioritizing local workers if available. This method
    * initializes the list of worker addresses if it has not been initialized.
    *
    * @return a file system worker client to a worker in the system
    * @throws IOException if an error occurs getting the list of workers in the system
    */
-  public FileSystemWorkerClient createWorkerClient() throws IOException {
+  public FileSystemWorkerClient createFileSystemWorkerClient() throws IOException {
     WorkerNetAddress address;
-    synchronized (mWorkerAddressesLock) {
+    synchronized (this) {
       if (mWorkerAddresses == null) {
         mWorkerAddresses = getWorkerAddresses();
       }
       address = mWorkerAddresses.get(ThreadLocalRandom.current().nextInt(mWorkerAddresses.size()));
     }
+
+    InetSocketAddress rpcAddress = NetworkAddressUtils.getRpcPortSocketAddress(address);
+    if (!mFileSystemWorkerClientPools.containsKey(rpcAddress)) {
+      FileSystemWorkerThriftClientPool pool =
+          new FileSystemWorkerThriftClientPool(mParentSubject, rpcAddress,
+              Configuration.getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_SIZE_MAX),
+              Configuration.getLong(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_GC_THRESHOLD_MS));
+      if (mFileSystemWorkerClientPools.putIfAbsent(rpcAddress, pool) != null) {
+        pool.close();
+      }
+    }
+
+    if (!mFileSystemWorkerClientHeartbeatPools.containsKey(rpcAddress)) {
+      FileSystemWorkerThriftClientPool pool =
+          new FileSystemWorkerThriftClientPool(mParentSubject, rpcAddress,
+              Configuration.getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_SIZE_MAX),
+              Configuration.getLong(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_GC_THRESHOLD_MS));
+      if (mFileSystemWorkerClientHeartbeatPools.putIfAbsent(rpcAddress, pool) != null) {
+        pool.close();
+      }
+    }
```

Comment:
This kind of logic is repeated for all the different pools, is it possible to abstract it to a method?